StatusTest.checkCounters intermittently fails

## Environment Details
* Helidon Version: 2.5.3-SNAPSHOT, 4.0.0-SNAPSHOT

----------

## Problem Description

I have seen this intermittently fail on the pipeline:

```
java.lang.AssertionError: 
Diff in counter 4xx
Expected: is <1L>
     but: was <0L>
	at io.helidon.examples.se.httpstatuscount.StatusTest.checkCounters(StatusTest.java:125)
	at io.helidon.examples.se.httpstatuscount.StatusTest.checkAfterStatus(StatusTest.java:118)
	at io.helidon.examples.se.httpstatuscount.StatusTest.checkStatusMetrics(StatusTest.java:88)
```

--- a/examples/metrics/http-status-count-se/src/main/java/io/helidon/examples/se/httpstatuscount/HttpStatusMetricService.java
+++ b/examples/metrics/http-status-count-se/src/main/java/io/helidon/examples/se/httpstatuscount/HttpStatusMetricService.java
@@ -15,6 +15,8 @@
  */
 package io.helidon.examples.se.httpstatuscount;
 
+import java.util.concurrent.atomic.AtomicInteger;
+
 import io.helidon.metrics.api.RegistryFactory;
 import io.helidon.webserver.Routing;
 import io.helidon.webserver.ServerRequest;
@@ -43,6 +45,8 @@ public class HttpStatusMetricService implements Service {
 
     static final String STATUS_TAG_NAME = "range";
 
+    private static final AtomicInteger IN_PROGRESS = new AtomicInteger();
+
     private final Counter[] responseCounters = new Counter[6];
 
     static HttpStatusMetricService create() {
@@ -69,8 +73,14 @@ public class HttpStatusMetricService implements Service {
         rules.any(this::updateRange);
     }
 
+    // for testing
+    static boolean isInProgress() {
+        return IN_PROGRESS.get() != 0;
+    }
+
     // Edited to adopt Ciaran's fix later in the thread.
     private void updateRange(ServerRequest request, ServerResponse response) {
+        IN_PROGRESS.incrementAndGet();
         response.whenSent()
                 .thenAccept(this::logMetric);
         request.next();
@@ -81,5 +91,6 @@ public class HttpStatusMetricService implements Service {
         if (range > 0 && range < responseCounters.length) {
             responseCounters[range].inc();
         }
+        IN_PROGRESS.decrementAndGet();
     }
 }

--- a/examples/metrics/http-status-count-se/src/test/java/io/helidon/examples/se/httpstatuscount/StatusTest.java
+++ b/examples/metrics/http-status-count-se/src/test/java/io/helidon/examples/se/httpstatuscount/StatusTest.java
@@ -15,10 +15,12 @@
  */
 package io.helidon.examples.se.httpstatuscount;
 
+import java.time.Duration;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 
-import io.helidon.common.http.Http;
+import io.helidon.common.http.Http.ResponseStatus;
+import io.helidon.common.http.Http.Status;
 import io.helidon.common.http.MediaType;
 import io.helidon.config.Config;
 import io.helidon.media.jsonp.JsonpSupport;
@@ -37,28 +39,19 @@ import org.junit.jupiter.api.BeforeAll;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
+import static org.hamcrest.CoreMatchers.not;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.isEmptyString;
+import static org.junit.jupiter.api.Assertions.fail;
 
 public class StatusTest {
+    private static final Duration TIMEOUT = Duration.ofSeconds(10);
 
     private static WebServer webServer;
     private static WebClient webClient;
 
     private final Counter[] STATUS_COUNTERS = new Counter[6];
-
-    @BeforeAll
-    static void init() {
-        Routing.Builder routingBuilder = Main.createRouting(Config.create());
-        routingBuilder.register("/status", new StatusService());
-
-        webServer = Main.startServer(routingBuilder).await();
-
-        webClient = WebClient.builder()
-                .baseUri("http://localhost:" + webServer.port())
-                .addMediaSupport(JsonpSupport.create())
-                .build();
-    }
 
     @AfterAll
     public static void stopServer() throws Exception {
@@ -67,6 +60,19 @@ public class StatusTest {
                     .toCompletableFuture()
                     .get(10, TimeUnit.SECONDS);
         }
+    }
+
+    @BeforeAll
+    static void init() {
+        Routing.Builder routingBuilder = Main.createRouting(Config.create());
+        routingBuilder.register("/status", new StatusService());
+
+        webServer = Main.startServer(routingBuilder).await(TIMEOUT);
+
+        webClient = WebClient.builder()
+                .baseUri("http://localhost:" + webServer.port())
+                .addMediaSupport(JsonpSupport.create())
+                .build();
     }
 
     @BeforeEach
@@ -80,13 +86,13 @@ public class StatusTest {
 
     @Test
     void checkStatusMetrics() throws ExecutionException, InterruptedException {
-        checkAfterStatus(171);
-        checkAfterStatus(200);
-        checkAfterStatus(201);
-        checkAfterStatus(204);
-        checkAfterStatus(301);
-        checkAfterStatus(401);
-        checkAfterStatus(404);
+        checkAfterStatus(ResponseStatus.create(171));
+        checkAfterStatus(Status.OK_200);
+        checkAfterStatus(Status.CREATED_201);
+        checkAfterStatus(Status.NO_CONTENT_204);
+        checkAfterStatus(Status.MOVED_PERMANENTLY_301);
+        checkAfterStatus(Status.UNAUTHORIZED_401);
+        checkAfterStatus(Status.NOT_FOUND_404);
     }
 
     @Test
@@ -100,26 +106,41 @@ public class StatusTest {
                 .accept(MediaType.APPLICATION_JSON)
                 .request()
                 .get();
-        assertThat("Status of /greet", response.status().code(), is(Http.Status.OK_200.code()));
-        checkCounters(response.status().code(), before);
+        assertThat("Status of /greet", response.status(), is(Status.OK_200));
+        String entity = response.content().as(String.class).await(TIMEOUT);
+        assertThat(entity, not(isEmptyString()));
+        checkCounters(response.status(), before);
     }
 
-    void checkAfterStatus(int status) throws ExecutionException, InterruptedException {
+    void checkAfterStatus(ResponseStatus status) throws ExecutionException, InterruptedException {
         long[] before = new long[6];
         for (int i = 1; i < 6; i++) {
             before[i] = STATUS_COUNTERS[i].getCount();
         }
         WebClientResponse response = webClient.get()
-                .path("/status/" + status)
+                .path("/status/" + status.code())
                 .accept(MediaType.APPLICATION_JSON)
                 .request()
                 .get();
-        assertThat("Response status", response.status().code(), is(status));
+
+        assertThat("Response status", response.status().code(), is(status.code()));
+        String entity = response.content().as(String.class).await(TIMEOUT);
+
         checkCounters(status, before);
     }
 
-    private void checkCounters(int status, long[] before) {
-        int family = status / 100;
+    private void checkCounters(ResponseStatus status, long[] before) throws InterruptedException {
+        // first make sure we do not have a request in progress
+        long now = System.currentTimeMillis();
+
+        while (HttpStatusMetricService.isInProgress()) {
+            Thread.sleep(50);
+            if (System.currentTimeMillis() - now > 5000) {
+                fail("Timed out while waiting for monitoring to finish");
+            }
+        }
+
+        int family = status.code() / 100;
         for (int i = 1; i < 6; i++) {
             long expectedDiff = i == family ? 1 : 0;
             assertThat("Diff in counter " + family + "xx", STATUS_COUNTERS[i].getCount() - before[i], is(expectedDiff));

--- a/examples/metrics/http-status-count-se/src/test/resources/application.yaml
+++ b/examples/metrics/http-status-count-se/src/test/resources/application.yaml
@@ -15,7 +15,7 @@
 #
 
 server:
-  port: 8080
+  port: 0
   host: 0.0.0.0
 
 app:

